catalog/spatial/registry/registry-api-impl/src/main/java/org/codice/ddf/registry/api/impl/RegistryMetacardHandler.java

```diff
@@ -0,0 +1,147 @@
+ * Copyright (c) Codice Foundation
+ * <p>
+ * This is free software: you can redistribute it and/or modify it under the terms of the GNU Lesser
+ * General Public License as published by the Free Software Foundation, either version 3 of the
+ * License, or any later version.
+ * <p>
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details. A copy of the GNU Lesser General Public License
+ * is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.ddf.registry.api.impl;
+
+import java.io.Serializable;
+import java.security.PrivilegedActionException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.commons.collections.CollectionUtils;
+import org.codice.ddf.registry.common.RegistryConstants;
+import org.codice.ddf.registry.common.metacard.RegistryObjectMetacardType;
+import org.codice.ddf.registry.federationadmin.service.FederationAdminException;
+import org.codice.ddf.registry.federationadmin.service.FederationAdminService;
+import org.codice.ddf.security.common.Security;
+import org.osgi.service.event.Event;
+import org.osgi.service.event.EventHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import ddf.catalog.data.Metacard;
+import ddf.catalog.data.impl.AttributeImpl;
+
+public class RegistryMetacardHandler implements EventHandler {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(RegistryMetacardHandler.class);
+
+    private static final int SHUTDOWN_TIMEOUT_SECONDS = 60;
+
+    private static final String METACARD_PROPERTY = "ddf.catalog.event.metacard";
+
+    private static final String CREATED_TOPIC = "ddf/catalog/event/CREATED";
+
+    private static final String UPDATED_TOPIC = "ddf/catalog/event/UPDATED";
+
+    private static final String DELETED_TOPIC = "ddf/catalog/event/DELETED";
+
+    private final ExecutorService executor;
+
+    private final FederationAdminService federationAdminService;
+
+    public RegistryMetacardHandler(ExecutorService executor,
+            FederationAdminService federationAdminService) {
+        this.executor = executor;
+        this.federationAdminService = federationAdminService;
+    }
+
+    @Override
+    public void handleEvent(Event event) {
+        Metacard mcard = (Metacard) event.getProperty(METACARD_PROPERTY);
+        if (mcard == null || !mcard.getTags().contains(RegistryConstants.REGISTRY_TAG_INTERNAL)) {
+            return;
+        }
+        executor.execute(() -> processEvent(mcard, event.getTopic()));
+    }
+
+    private void processEvent(Metacard mcard, String topic) {
+        try {
+            Security.runAsAdminWithException(() -> {
```

Comment:
You should do the run with admin around each process method call instead of up here. At this level any and all events are going to require you to go off and get an assertion even if it isn't an event that you care about. This could negatively impact system performance.